docs/clamscan.1
.\" Manual page created by Tomasz Kojm, 14/15 IV 2002
.TH "clamscan" "1" "October 1, 2002" "Tomasz Kojm" "Clam AntiVirus"
.SH "NAME"
.LP 
clamscan \- scan files/directories against viruses
.SH "SYNOPSIS"
.LP 
clamscan [options] [file/directory/\-]
.SH "DESCRIPTION"
.LP 
clamscan is a console anti\-virus scanner. It's a part of the Clam AntiVirus package.
.SH "OPTIONS"
.LP 

.TP 
\fB\-h, \-\-help\fR
Output help information and exit.
.TP 
\fB\-V, \-\-version\fR
Print the version number and exit.
.TP 
\fB\-v, \-\-verbose\fR
Be verbose. This option causes clamscan to print many additional informations.
.TP 
\fB\-\-quiet\fR
Be quiet \- output only error messages.
.TP 
\fB\-\-stdout\fR
Write all messages to the standard output (stdout), instead of the standard error output (stderr). The buffer is automatically flushed (stdout unlike stderr is buffered), to avoid a race conditions on output.
.TP 
\fB\-d FILE/DIR, \-\-database=FILE/DIR\fR
Load the virus database from FILE or load all .db and .db2 files from DIR.
.TP 
\fB\-l FILE, \-\-log=FILE\fR
Save the scan report to FILE. By default the log isn't saved.
.TP 
\fB\-\-log\-verbose\fR
Save an additional information (mostly needless) to the report file.
.TP 
\fB\-\-tempdir=DIRECTORY\fR
Create temporary files in DIRECTORY. Directory must be writeable for the 'clamav' user or unprivileged user running clamscan.
.TP 
\fB\-r, \-\-recursive\fR
Scan directories recursively. All the files in the given directory will be scanned. \fBThis option affects on scanning subdirectories in archives, if you are using unpackers just turn it on.\fR
.TP 
\fB\-\-disable\-summary\fR
Disable summary printing at the end of the scanning.
.TP 
\fB\-\-exclude=PATT\fR
Don't scan file names containing PATT. It may be used multiple times.
.TP 
\fB\-\-include=PATT\fR
Only scan file names containing PATT. It may be used multiple times.
.TP 
\fB\-\-mbox\fR
Enable scanning of mbox and Maildir files. (also treat stdin as a mailbox, for backward compatibility).
\fB\-i, \-\-infected\fR
Print infected files only.
.TP 
\fB\-\-remove\fR
Remove infected files. \fBBe careful.\fR
.TP
\fB\-\-move=DIRECTORY\fR
Move infected files into DIRECTORY. Directory must be writeable for the 'clamav' user or unprivileged user running clamscan.
.TP 
\fB\-\-force\fR
With this option you may ignore some errors. \fBBe careful.\fR
.TP 
EXTRACTION OPTIONS:
.TP 
\fB\-\-disable\-archive\fR
Disable archive support built in libclamav.
.TP 
\fB\-\-max\-files=#n\fR
Extract first #n files from each archive. This option should be used to protect your system against DoS attacks. Note: always few files more are extracted due to the limit implementation (expecially when archive contains small files). All file types are counted.
.TP 
\fB\-\-max\-space=#n\fR
Extract first #n kilobytes from each archive. You may give the number in megabytes in format xM or xm, where x is a number.
.TP 
\fB\-\-max\-recursion=#n\fR
Set the archive recursion level.
.TP 
\fB\-\-unzip[=FULLPATH]\fR
With this option all zipped files are scanned. If you turn on this option without the argument, unzip program will be searched in $PATH. If unzip is outside $PATH, please supply it with =pathname. Remember about '=' between option and argument. (This is needed due to the implementation of getopt_long() and optional arguments to long options).
.TP 
\fB\-\-unrar[=FULLPATH]\fR
Supports .rar files.
.TP 
\fB\-\-unace[=FULLPATH]\fR
Supports .ace files.
.TP 
\fB\-\-unarj[=FULLPATH]\fR
Supports .arj files.
.TP 
\fB\-\-unzoo[=FULLPATH]\fR
Supports .zoo files.
.TP 
\fB\-\-lha[=FULLPATH]\fR
Supports .lzh files.
.TP 
\fB\-\-jar[=FULLPATH]\fR
clamscan uses unzip for .jar files, so optionally give a full path to unzip.
.TP 
\fB\-\-deb[=FULLPATH]\fR
This option supports debian binary packages. Implies \-\-tgz, but doesn't conflict with \-\-tgz=FULLPATH. It requires ar utility.
.TP 
\fB\-\-tar[=FULLPATH]\fR
This option supports non\-compressed archives.
.TP 
\fB\-\-tgz[=FULLPATH]\fR
This option supports tar.gz and .tgz files. You need GNU tar, on non\-Linux system you probably have it as gtar. If it's in $PATH, please use \-\-tgz=gtar in other case please use a full pathname.
.SH "EXAMPLES"
.LP 
.TP 
(0) To scan one file:

\fBclamscan file\fR
.TP 
(1) To scan current directory:

\fBclamscan\fR
.TP 
(2) To scan all files in /home:

\fBclamscan \-r /home\fR
.TP 
(3) Load the database from another file and limit the disk usage to 50 Mb:

\fBclamscan \-d /tmp/newclamdb \-\-max-space=50m \-r /tmp\fR
.TP 
(4) Scan a data stream:

\fBcat testfile | clamscan \-\fR
.TP 
(5) Scan the mail spool directory:

\fBclamscan \-r \-\-mbox /var/spool/mail\fR
.SH "RETURN CODES"
.LP 
Note: some return codes may appear only in a one file mode (clamscan is started with file argument). These one are marked with \fB(ofm)\fR.

0 : No virus found.
.TP 
1 : Virus(es) found.
.TP 
40: Unknown option passed.
.TP 
50: The database can't be initialized.
.TP 
51: Wrong number of threads given.
.TP 
52: Not supported file type.
.TP 
53: Can't open directory.
.TP 
54: Can't open file. (ofm)
.TP 
55: Error reading file. (ofm)
.TP 
56: Can't stat input file / directory.
.TP 
57: Can't get absolute pathname of current working directory.
.TP 
58: I/O error, check your filesystem.
.TP 
59: Can't get information about current user from /etc/passwd.
.TP 
60: Can't get information about user 'clamav' (default name) from /etc/passwd.
.TP 
61: Can't fork.
.TP 
63: Can't create temporary files/directories (check permissions).
.TP 
64: Can't write to temporary directory, please specify another one.
.TP 
70: Can't allocate and clear memory (calloc).
.TP 
71: Can't allocate memory (malloc).
.SH "CREDITS"
Please check the full documentation for credits.
.SH "AUTHOR"
.LP 
Tomasz Kojm <zolw@konarski.edu.pl>
.SH "SEE ALSO"
.LP 
clamd(8), clamdscan(1), freshclam(1), sigtool(1), clamav.conf(5)


NEWS

0.100.0-beta
------

Welcome to ClamAV 0.100.0.  0.100.0 is under active development.

It includes many code submissions from the ClamAV community:
    - Interfaces to the Prelude SIEM open source package for collecting
      ClamAV virus events.
    - Visual Studio 2015 for building Microsoft Windows binaries.
    - Support libmspack internal code or as a shared object library.
      The internal library is the default and contains additional
      integrity checks.
    - Linking with openssl 1.1.0.
    - Deprecation of the AllowSupplementaryGroups parameter statement
      in clamd, clamav-milter, and freshclam. Use of supplementary
      is now in effect by default.
    - Numerous bug fixes, typo corrections, and compiler warning fixes.

Additionally, we have introduced important changes and new features in
ClamAV 0.100, including:
    - Deprecating internal LLVM code support. The configure script has changed
      to search the system for an installed instance of the LLVM development
      libraries, and to otherwise use the bytecode interpreter for ClamAV
      bytecode signatures. To use the LLVM Just-In-Time compiler for
      executing bytecode signatures, please ensure that the LLVM development
      package at version 3.6 or lower is installed. Using the deprecated LLVM
      code is possible with the command: './configure --with-system-llvm=no',
      but it no longer compiles on all platforms.
    - Compute and check PE import table hash (a.k.a. "imphash") signatures.
    - Support file property collection and analysis for MHTML files.
    - Raw scanning of PostScript files.
    - Fix clamsubmit to use the new virus and false positive submission web
      interface.
    - Optionally, flag files with the virus "Heuristic.Limits.Exceeded" when
      size limitations are exceeded.
    - Improve decoders for PDF files.

The ClamAV team thanks the following individuals for their code submissions:

Andreas Schulze
Anthony Chan
Bill Parker
Chris Miserva
Daniel J. Luke
Georgy Salnikov
Jonas Zaddach
Keith Jones
Marc Deslauriers
Mark Allan
Matthew Boedicker
Michael Pelletier
Ningirsu
Sebastian Andrzej Siewior
Stephen Welker

--
The ClamAV team (http://www.clamav.net/about.html#credits)


clamav-devel/clamscan/clamscan.c
/*
 *  Copyright (C) 2002 - 2006 Tomasz Kojm <tkojm@clamav.net>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 *  MA 02110-1301, USA.
 */

#if HAVE_CONFIG_H
#include "clamav-config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/time.h>
#include <time.h>
#ifdef C_LINUX
#include <sys/resource.h>
#endif

#include "clamscan_opt.h"
#include "others.h"
#include "global.h"
#include "manager.h"
#include "treewalk.h"

#include "shared/misc.h"
#include "shared/output.h"
#include "shared/options.h"

void help(void);

struct s_info info;
short recursion = 0, printinfected = 0, bell = 0;

int main(int argc, char **argv)
{
	int ds, dms, ret;
	double mb;
	struct timeval t1, t2;
	struct timezone tz;
	struct optstruct *opt;
	const char *pt;


    opt = opt_parse(argc, argv, clamscan_shortopt, clamscan_longopt, NULL);
    if(!opt) {
	mprintf("!Can't parse the command line\n");
	return 40;
    }

    if(opt_check(opt, "verbose")) {
	mprintf_verbose = 1;
	logg_verbose = 1;
    }

    if(opt_check(opt, "quiet"))
	mprintf_quiet = 1;

    if(opt_check(opt, "stdout"))
	mprintf_stdout = 1;


    if(opt_check(opt, "debug")) {
#if defined(C_LINUX)
	    /* njh@bandsman.co.uk: create a dump if needed */
	    struct rlimit rlim;

	rlim.rlim_cur = rlim.rlim_max = RLIM_INFINITY;
	if(setrlimit(RLIMIT_CORE, &rlim) < 0)
	    perror("setrlimit");
#endif
	cl_debug(); /* enable debug messages */
    }

    if(opt_check(opt, "version")) {
	opt_free(opt);
	print_version();
	return 0;
    }

    if(opt_check(opt, "help")) {
	opt_free(opt);
    	help();
	return 0;
    }

    if(opt_check(opt, "recursive"))
	recursion = 1;

    if(opt_check(opt, "infected"))
	printinfected = 1;

    if(opt_check(opt, "bell"))
	bell = 1;

    if(opt_check(opt, "tempdir"))
	cl_settempdir(opt_arg(opt, "tempdir"), 0);

    if(opt_check(opt, "leave-temps"))
	cl_settempdir(NULL, 1);

    /* initialize logger */
    if(opt_check(opt, "log")) {
	logg_file = opt_arg(opt, "log");
	if(logg("#\n-------------------------------------------------------------------------------\n\n")) {
	    mprintf("!Problem with internal logger.\n");
	    opt_free(opt);
	    return 62;
	}
    } else 
	logg_file = NULL;


    /* validate some numerical options */

    if(opt_check(opt, "max-space")) {
	pt = opt_arg(opt, "max-space");
	if(!strchr(pt, 'M') && !strchr(pt, 'm')) {
	    if(!isnumb(pt)) {
		logg("!--max-space requires a natural number\n");
		opt_free(opt);
		return 40;
	    }
	}
    }

    if(opt_check(opt, "max-files")) {
	if(!isnumb(opt_arg(opt, "max-files"))) {
	    logg("!--max-files requires a natural number\n");
	    opt_free(opt);
	    return 40;
	}
    }

    if(opt_check(opt, "max-recursion")) {
	if(!isnumb(opt_arg(opt, "max-recursion"))) {
	    logg("!--max-recursion requires a natural number\n");
	    opt_free(opt);
	    return 40;
	}
    }

    if(opt_check(opt, "max-mail-recursion")) {
	if(!isnumb(opt_arg(opt, "max-mail-recursion"))) {
	    logg("!--max-mail-recursion requires a natural number\n");
	    opt_free(opt);
	    return 40;
	}
    }

    if(opt_check(opt, "max-dir-recursion")) {
	if(!isnumb(opt_arg(opt, "max-dir-recursion"))) {
	    logg("!--max-dir-recursion requires a natural number\n");
	    opt_free(opt);
	    return 40;
	}
    }

    if(opt_check(opt, "max-ratio")) {
	if(!isnumb(opt_arg(opt, "max-ratio"))) {
	    logg("!--max-ratio requires a natural number\n");
	    opt_free(opt);
	    return 40;
	}
    }

    memset(&info, 0, sizeof(struct s_info));

    gettimeofday(&t1, &tz);
    ret = scanmanager(opt);

    if(!opt_check(opt, "disable-summary") && !opt_check(opt, "no-summary")) {
	gettimeofday(&t2, &tz);
	ds = t2.tv_sec - t1.tv_sec;
	dms = t2.tv_usec - t1.tv_usec;
	ds -= (dms < 0) ? (1):(0);
	dms += (dms < 0) ? (1000000):(0);
	logg("\n----------- SCAN SUMMARY -----------\n");
	logg("Known viruses: %u\n", info.sigs);
	if(opt_check(opt, "ncore"))
	    logg("Engine version: %s [ncore]\n", cl_retver());
	else
	    logg("Engine version: %s\n", cl_retver());
	logg("Scanned directories: %u\n", info.dirs);
	logg("Scanned files: %u\n", info.files);
	logg("Infected files: %u\n", info.ifiles);
	if(info.notremoved) {
	    logg("Not removed: %u\n", info.notremoved);
	}
	if(info.notmoved) {
	    logg("Not %s: %u\n", opt_check(opt, "copy") ? "moved" : "copied", info.notmoved);
	}
	mb = info.blocks * (CL_COUNT_PRECISION / 1024) / 1024.0;
	logg("Data scanned: %2.2lf MB\n", mb);
	logg("Time: %u.%3.3u sec (%u m %u s)\n", ds, dms/1000, ds/60, ds%60);
    }

    opt_free(opt);
    return ret;
}

void help(void)
{

    mprintf_stdout = 1;

    mprintf("\n");
    mprintf("                       Clam AntiVirus Scanner "VERSION"\n");
    mprintf("    (C) 2002 - 2005 ClamAV Team - http://www.clamav.net/team.html\n\n");

    mprintf("    --help                -h             Print this help screen\n");
    mprintf("    --version             -V             Print version number\n");
    mprintf("    --verbose             -v             Be verbose\n");
    mprintf("    --debug                              Enable libclamav's debug messages\n");
    mprintf("    --quiet                              Only output error messages\n");
    mprintf("    --stdout                             Write to stdout instead of stderr\n");
    mprintf("    --no-summary                         Disable summary at end of scanning\n");
    mprintf("    --infected            -i             Only print infected files\n");
    mprintf("    --bell                               Sound bell on virus detection\n");

    mprintf("\n");
    mprintf("    --tempdir=DIRECTORY                  Create temporary files in DIRECTORY\n");
    mprintf("    --leave-temps                        Do not remove temporary files\n");
    mprintf("    --database=FILE/DIR   -d FILE/DIR    Load virus database from FILE or load\n");
    mprintf("                                         all .cvd and .db[2] files from DIR\n");
    mprintf("    --log=FILE            -l FILE        Save scan report to FILE\n");
    mprintf("    --recursive           -r             Scan subdirectories recursively\n");
    mprintf("    --remove                             Remove infected files. Be careful!\n");
    mprintf("    --move=DIRECTORY                     Move infected files into DIRECTORY\n");
    mprintf("    --copy=DIRECTORY                     Copy infected files into DIRECTORY\n");
#ifdef HAVE_REGEX_H
    mprintf("    --exclude=REGEX                      Don't scan file names matching REGEX\n");
    mprintf("    --exclude-dir=REGEX                  Don't scan directories matching REGEX\n");
    mprintf("    --include=REGEX                      Only scan file names matching REGEX\n");
    mprintf("    --include-dir=REGEX                  Only scan directories matching REGEX\n");
#else
    mprintf("    --exclude=PATT                       Don't scan file names containing PATT\n");
    mprintf("    --exclude-dir=PATT                   Don't scan directories containing PATT\n");
    mprintf("    --include=PATT                       Only scan file names containing PATT\n");
    mprintf("    --include-dir=PATT                   Only scan directories containing PATT\n");
#endif
#ifdef HAVE_NCORE
    mprintf("\n    --ncore                            Use hardware acceleration\n");
#endif
    mprintf("\n");
    mprintf("    --no-mail                            Disable mail file support\n");
    mprintf("    --no-phishing-sigs                   Disable signature-based phishing detection\n");
#ifdef CL_EXPERIMENTAL
    mprintf("    --no-phishing-scan-urls              Disable url-based phishing detection\n");
    mprintf("    --no-phishing-restrictedscan         Enable phishing detection for all domains (might lead to false positives!)\n");
    mprintf("    --phishing-ssl                       Always block SSL mismatches in URLs (phishing module)\n");
    mprintf("    --phishing-cloak                     Always block cloaked URLs (phishing module)\n");
#endif
    mprintf("    --no-algorithmic                     Disable algorithmic detection\n");
    mprintf("    --no-pe                              Disable PE analysis\n");
    mprintf("    --no-elf                             Disable ELF support\n");
    mprintf("    --no-ole2                            Disable OLE2 support\n");
    mprintf("    --no-html                            Disable HTML support\n");
    mprintf("    --no-archive                         Disable libclamav archive support\n");
    mprintf("    --detect-broken                      Try to detect broken executable files\n");
    mprintf("    --block-encrypted                    Block encrypted archives\n");
    mprintf("    --block-max                          Block archives that exceed limits\n");
#ifdef WITH_CURL
    mprintf("    --mail-follow-urls                   Download and scan URLs\n");
#endif

    mprintf("\n");
    mprintf("    --max-space=#n                       Only extract first #n kilobytes from\n");
    mprintf("                                         archived files\n");
    mprintf("    --max-files=#n                       Only extract first #n files from\n");
    mprintf("                                         archives\n");
    mprintf("    --max-ratio=#n                       Maximum compression ratio limit\n");
    mprintf("    --max-recursion=#n                   Maximum archive recursion level\n");
    mprintf("    --max-dir-recursion=#n               Maximum directory recursion level\n");
    mprintf("    --max-mail-recursion=#n              Maximum mail recursion level\n");
    mprintf("    --unzip[=FULLPATH]                   Enable support for .zip files\n");
    mprintf("    --unrar[=FULLPATH]                   Enable support for .rar files\n");
    mprintf("    --arj[=FULLPATH]                     Enable support for .arj files\n");
    mprintf("    --unzoo[=FULLPATH]                   Enable support for .zoo files\n");
    mprintf("    --lha[=FULLPATH]                     Enable support for .lha files\n");
    mprintf("    --jar[=FULLPATH]                     Enable support for .jar files\n");
    mprintf("    --tar[=FULLPATH]                     Enable support for .tar files\n");
    mprintf("    --deb[=FULLPATH to ar]               Enable support for .deb files\n");
    mprintf("    --tgz[=FULLPATH]                     Enable support for .tar.gz, .tgz files\n\n");
}
